L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,465 @@
+* A concrete example of how code can benefit from Reactive Streams is [Netflix's Hystrix library](https://github.com/Netflix/Hystrix/wiki/How-it-Works), which helps provide latency and fault tolerance for microservices.
+
+The main drawbacks of this proposal are:
+
+* It is easy to use only after you have spent some time learning Reactive Streams' concepts: You must change mindset to thinking about data flows.
+* There are features in gRPC that don't obviously fit Reactive Streams, for example metadata.
+* It is a large change: Broad adoption will require a lot of education and a substantial change of mindset for programmers that use it. Netflix has adopted Rx at a large scale and their engineers report that it was worth it for them, but it required a fairly large educational effort.
+* This proposal is based on the assumption that the current async C++ gRPC API is too low level to be user friendly for typical application development. If people do not agree with this, it is pointless.
+
+
+### Why Reactive Streams?
+
+Although Reactive Streams fits gRPC remarkably well, it can be difficult to grasp. What the proposed API achieves could be done without Reactive Streams. Why use Reactive Streams and not something tailored to gRPC?
+
+The reason is that Reactive Streams is a well understood paradigm with a lot of momentum:
+
+It is being used extensively by [Netflix](https://medium.com/netflix-techblog/reactive-programming-in-the-netflix-api-with-rxjava-7811c3a1496a), [Akka](http://doc.akka.io/docs/akka/2.5.4/scala/stream/stream-introduction.html), Github and many other companies. The upcoming version 5 of the widely used Spring framework [is based on Reactive Streams](https://spring.io/blog/2016/07/28/reactive-programming-with-spring-5-0-m1). A precursor to Reactive Streams is in [LINQ in C#](https://msdn.microsoft.com/en-us/library/hh242977(v=vs.103).aspx). Reactive Streams [is being standardized in Java 9](https://community.oracle.com/docs/DOC-1006738). With [ReactiveX](http://reactivex.io/languages.html) this programming model is available in lots of programming languages (albeit not always with backpressure support).
+
+There is a lot of material on the Internet that covers Reactive Streams from many different angles. [ReactiveX](http://reactivex.io/) has lots of good documentation and [Marble diagrams](http://rxmarbles.com/) are a wonderful way to describe what Rx operators do.
+
+
+## Implementation
+
+A prototype of what is suggested in this RFC can be found [here](https://github.com/per-gron/shuriken/tree/master/src/rs-grpc). It implements the most important features, but it is still quite incomplete.
+
+Implementation of this RFC would involve:
+
+* Prepare the [rs](https://github.com/per-gron/shuriken/tree/master/src/rs) library so that it can be accepted as gRPC code:
+  * Backport to C++11.
+  * Remove usage of `std::exception_ptr` and make it possible to use `grpc::Status` instead.
```

Comment:
@ctiller you said (starting a new thread because the other one is hidden by the Github interface)

> grpc also has a generic container for error handling (grpc::Status). It's hard to justify introducing a wrinkle that forces conversion.

Would you use `grpc::Status` for non-gRPC errors also, like HTTP errors or Cassandra errors? A strength of Reactive Streams is that it provides a uniform API style. Unless you would advocate for wrapping non-gRPC errors in `grpc::Status`, I think there might be a good case for wrapping errors, to be able to handle different kinds.

The syntactic overhead does not have to be that large. In the current version of rs, you usually catch errors like this: `Catch([](std::exception_ptr error) { ... })`. Instead of this, `Catch` could allow you to specify what you want to catch:

```
Catch<grpc::Status>([](grpc::Status status) { ... })
```

That way, you could also catch your `CassandraError`s or `HttpError`s or whatever, like you would with normal `try`/`catch` statements. The only extra code the user has to write is `<grpc::Status>`. Since the benefit of doing it like this is to not force users to encode non-gRPC errors in `grpc::Status`, it could be worth it.

This type of `try`/`catch` style with type matching could be done using `std::exception_ptr` but it could also without too much effort be done with an rs wrapper type.